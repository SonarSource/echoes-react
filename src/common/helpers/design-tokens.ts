/*
 * Echoes React
 * Copyright (C) 2023-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import designTokensColors from '~generated/design-tokens-colors.json';

export type DesignTokens = keyof typeof designTokensColors;

export type DesignTokensColors = {
  [K in DesignTokens]: K extends `echoes-color-${string}` ? K : never;
}[DesignTokens];
