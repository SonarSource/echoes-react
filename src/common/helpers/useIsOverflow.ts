/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { DependencyList, MutableRefObject, useLayoutEffect, useState } from 'react';

/**
 * This hook checks if the content of a given ref is overflowing its container horizontally.
 * It returns a boolean indicating whether the content is overflowing.
 *
 * @param forwardedRef - The ref to check for overflow.
 * @param deps - Optional dependencies array to re-compute the overflow state.
 */
export function useIsOverflow(
  forwardedRef: MutableRefObject<HTMLElement | null>,
  deps: DependencyList = [],
) {
  const [isOverflow, setIsOverflow] = useState<boolean | undefined>(undefined);

  useLayoutEffect(() => {
    if (forwardedRef.current) {
      setIsOverflow(forwardedRef.current.scrollWidth > forwardedRef.current.offsetWidth);
    }
    // We want to be able to pass an array of dependencies to re-compute the overflow state
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [forwardedRef, ...deps]);

  return [isOverflow] as const;
}
