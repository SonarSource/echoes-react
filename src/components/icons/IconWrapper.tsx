/*
 * Echoes React
 * Copyright (C) 2023-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import styled from '@emotion/styled';

export type IconProps = Omit<JSX.IntrinsicElements['span'], 'children'>;

export function IconWrapper(props: Readonly<IconProps & { children: React.ReactNode }>) {
  const { children, className, ...rest } = props;

  return (
    <StyledSpan className={className} {...rest}>
      {children}
    </StyledSpan>
  );
}

const StyledSpan = styled.span`
  font-family: 'Material Symbols Rounded';
  font-size: 1.2em;
  font-weight: normal;
  font-style: normal;
  line-height: 1;
  letter-spacing: normal;
  text-transform: none;
  white-space: nowrap;
  word-wrap: normal;
  direction: ltr;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  text-rendering: optimizeLegibility;
`;
