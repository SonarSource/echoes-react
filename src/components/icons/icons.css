@font-face {
  font-family: 'Material Symbols Rounded';
  font-style: normal;
  font-weight: 100 700;
  font-display: block;
  src: url('material-symbols/material-symbols-rounded.woff2') format('woff2');
}

@font-face {
  font-family: 'Echoes Icons';
  font-style: normal;
  font-weight: 400;
  font-display: auto;
  src: url('./echoes-icons.woff2') format('woff2');
}
