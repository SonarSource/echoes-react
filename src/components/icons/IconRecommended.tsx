/*
 * Echoes React
 * Copyright (C) 2023-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { forwardRef } from 'react';
import { IconFilledProps, IconMaterialWrapper } from './IconWrapper';

export const IconRecommended = forwardRef<HTMLSpanElement, IconFilledProps>((props, ref) => {
  // This is Material Symbols' "new_releases" icon
  return (
    <IconMaterialWrapper {...props} ref={ref}>
      &#xE031;
    </IconMaterialWrapper>
  );
});
IconRecommended.displayName = 'IconRecommended';
