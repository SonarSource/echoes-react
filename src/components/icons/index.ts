/*
 * Echoes React
 * Copyright (C) 2023-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import './icons.css';

export type { IconProps } from './IconWrapper';

export { IconArrowDownRight } from './IconArrowDownRight';
export { IconArrowLeft } from './IconArrowLeft';
export { IconArrowRight } from './IconArrowRight';
export { IconArrowUpRight } from './IconArrowUpRight';
export { IconBell } from './IconBell';
export { IconBranch } from './IconBranch';
export { IconBug } from './IconBug';
export { IconCalendar } from './IconCalendar';
export { IconCheck } from './IconCheck';
export { IconCheckCircle } from './IconCheckCircle';
export { IconChevronDown } from './IconChevronDown';
export { IconClock } from './IconClock';
export { IconCodeSmell } from './IconCodeSmell';
export { IconCollapse } from './IconCollapse';
export { IconComment } from './IconComment';
export { IconCopy } from './IconCopy';
export { IconDash } from './IconDash';
export { IconDirectory } from './IconDirectory';
export { IconDot } from './IconDot';
export { IconDownload } from './IconDownload';
export { IconEqual } from './IconEqual';
export { IconError } from './IconError';
export { IconExpand } from './IconExpand';
export { IconFile } from './IconFile';
export { IconFileCode } from './IconFileCode';
export { IconFilter } from './IconFilter';
export { IconGear } from './IconGear';
export { IconGitBranch } from './IconGitBranch';
export { IconHome } from './IconHome';
export { IconInfo } from './IconInfo';
export { IconInheritance } from './IconInheritance';
export { IconLink } from './IconLink';
export { IconLinkExternal } from './IconLinkExternal';
export { IconLock } from './IconLock';
export { IconMegaphone } from './IconMegaphone';
export { IconMoreVertical } from './IconMoreVertical';
export { IconOverview } from './IconOverview';
export { IconPeople } from './IconPeople';
export { IconPin } from './IconPin';
export { IconPlus } from './IconPlus';
export { IconProject } from './IconProject';
export { IconPullrequest } from './IconPullRequest';
export { IconQuestionMark } from './IconQuestionMark';
export { IconRecommended } from './IconRecommended';
export { IconRocket } from './IconRocket';
export { IconSearch } from './IconSearch';
export { IconSecurityFinding } from './IconSecurityFinding';
export { IconSlideshow } from './IconSlideshow';
export { IconStar } from './IconStar';
export { IconStatusConfirmed } from './IconStatusConfirmed';
export { IconStatusOpen } from './IconStatusOpen';
export { IconStatusReopened } from './IconStatusReopened';
export { IconStatusResolved } from './IconStatusResolved';
export { IconTarget } from './IconTarget';
export { IconUnfold } from './IconUnfold';
export { IconUnfoldDown } from './IconUnfoldDown';
export { IconUnfoldUp } from './IconUnfoldUp';
export { IconVulnerability } from './IconVulnerability';
export { IconWarning } from './IconWarning';
export { IconX } from './IconX';
