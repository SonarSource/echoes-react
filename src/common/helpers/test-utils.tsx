/*
 * Echoes React
 * Copyright (C) 2023-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { RenderOptions, render as rtlRender } from '@testing-library/react';
import userEvent, { Options as UserEventsOptions } from '@testing-library/user-event';
import React, { ComponentProps, PropsWithChildren } from 'react';
import { IntlProvider } from 'react-intl';
import { PropsWithLabels } from '~types/utils';
import { TooltipProvider } from '../../components/tooltip';

export function render(
  ui: React.ReactElement,
  options?: RenderOptions,
  userEventOptions?: UserEventsOptions,
) {
  return {
    ...rtlRender(ui, { wrapper: ContextWrapper, ...options }),
    user: userEvent.setup(userEventOptions),
  };
}

function ContextWrapper({ children }: PropsWithChildren<{}>) {
  return (
    <IntlProvider defaultLocale="en-us" locale="en-us">
      <TooltipProvider delayDuration={0}>{children}</TooltipProvider>
    </IntlProvider>
  );
}

export type OmitPropsWithLabels<T extends React.JSXElementConstructor<any>> = Partial<
  Omit<ComponentProps<T>, keyof PropsWithLabels<{}>>
> &
  PropsWithLabels<{}>;
