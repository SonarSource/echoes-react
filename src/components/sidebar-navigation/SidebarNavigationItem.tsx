/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { css } from '@emotion/react';
import styled from '@emotion/styled';
import classNames from 'classnames';
import { forwardRef, ForwardRefExoticComponent, MouseEventHandler } from 'react';
import { NavLinkBase, NavLinkBaseProps } from '~common/components/NavLinkBase';
import { TextNode } from '~types/utils';
import { cssVar } from '~utils/design-tokens';
import { IconProps } from '../icons';
import { Tooltip } from '../tooltip';
import {
  itemIconStyles,
  sidebarNavigationBaseItemStyles,
  SidebarNavigationItemLabel,
} from './SidebarNavigationItemStyles';

export interface SidebarNavigationItemProps
  extends Pick<NavLinkBaseProps, 'isMatchingFullPath' | 'enableOpenInNewTab' | 'to'> {
  ariaLabel?: string;
  /**
   * The label of the SidebarNavigationItem.
   * It can be a string or a JSX.Element, in the case of a JSX.Element it should not be wrapped in
   * a `<Text>` component, the SidebarNavigationItem already handles the typography styling for you.
   */
  children: TextNode;
  className?: string;
  /**
   * Whether to display the tooltip on the item or not.
   * By default the tooltip is disabled, it should only be enabled if you expect the content to be ellipsed.
   */
  enableTooltip?: boolean;
  /**
   * Control whether the SidebarNavigationItem is active or not.
   * If true, the item will have a different style to indicate it is active.
   *
   * By default this behavior is handled by the underlying react-router's NavLink component,
   * overriding this is only needed for complex scenarios.
   */
  isActive?: boolean;
  /**
   * The onClick handler for the SidebarNavigationItem.
   */
  onClick?: MouseEventHandler<HTMLAnchorElement>;
  /**
   * The icon component to display at the start of the SidebarNavigationItem.
   * Must be an Echoes Icon component.
   */
  Icon?: ForwardRefExoticComponent<IconProps & React.RefAttributes<HTMLSpanElement>>;
}

export const SidebarNavigationItem = forwardRef<HTMLAnchorElement, SidebarNavigationItemProps>(
  (props, ref) => {
    const { children, className, enableTooltip, Icon, isActive = false, ...htmlProps } = props;

    return (
      <Tooltip content={enableTooltip ? children : undefined} side="right">
        <NavigationItem
          {...htmlProps}
          className={classNames({ active: isActive }, className)}
          ref={ref}>
          {Icon ? <Icon css={navigationItemIconStyles} /> : undefined}

          <SidebarNavigationItemLabel>{children}</SidebarNavigationItemLabel>
        </NavigationItem>
      </Tooltip>
    );
  },
);

SidebarNavigationItem.displayName = 'SidebarNavigationItem';

const NavigationItem = styled(NavLinkBase)`
  ${sidebarNavigationBaseItemStyles}

  // When the item is inside an accordion, the display value change based on the state of the accordion, hidding the items when collapsed
  // This css property is set by the SidebarNavigationAccordionItem component
  // Fallback to flex if not inside an accordion
  display: var(--sidebar-navigation-accordion-children-display, flex);

  &:active,
  &.active {
    // Always display the item when active even if behind a collapsed accordion, this override the previously set display value from the css property
    display: flex;

    background-color: ${cssVar('color-background-selected-weak-default')};
    color: ${cssVar('color-text-accent')};
    font: ${cssVar('typography-text-default-semi-bold')};

    &:hover {
      background-color: ${cssVar('color-background-selected-weak-hover')};
    }
  }
`;
NavigationItem.displayName = 'NavigationItem';

const navigationItemIconStyles = css`
  ${itemIconStyles}

  ${NavigationItem}.active > &,
  ${NavigationItem}:active > & {
    color: ${cssVar('color-icon-accent')};
  }
`;
