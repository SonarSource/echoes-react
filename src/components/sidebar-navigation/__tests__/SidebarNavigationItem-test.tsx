/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { screen } from '@testing-library/react';
import { renderWithMemoryRouter } from '~common/helpers/test-utils';
import { useIsOverflow } from '~common/helpers/useIsOverflow';
import { IconBranch } from '../../icons';
import { SidebarNavigationItem, SidebarNavigationItemProps } from '../SidebarNavigationItem';

// Mock the useIsOverflow hook
jest.mock('~common/helpers/useIsOverflow', () => ({
  useIsOverflow: jest.fn().mockReturnValue([false]),
}));

it('should render with an icon', () => {
  setupSidebarNavigationItem({ Icon: IconBranch });

  expect(screen.getByRole('link')).toMatchSnapshot();
});

it('should apply active class when isActive is true', () => {
  setupSidebarNavigationItem({ isActive: true });

  expect(screen.getByRole('link')).toHaveClass('active');
});

it('should handle onClick events', async () => {
  const handleClick = jest.fn();
  const { user } = setupSidebarNavigationItem({ onClick: handleClick });

  await user.click(screen.getByRole('link'));
  expect(handleClick).toHaveBeenCalledTimes(1);
});

describe('Ellipsis behavior', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should show tooltip when content is overflowing', async () => {
    jest.mocked(useIsOverflow).mockReturnValueOnce([true]);
    const { user } = setupSidebarNavigationItem();

    await user.hover(screen.getByRole('link'));
    const tooltip = await screen.findByRole('tooltip');
    expect(tooltip).toBeInTheDocument();
    expect(tooltip).toHaveTextContent('Test Item');
  });

  it('should not show tooltip when content is not overflowing', async () => {
    const { user } = setupSidebarNavigationItem();

    await user.hover(screen.getByRole('link'));
    expect(screen.queryByRole('tooltip')).not.toBeInTheDocument();
  });
});

describe('Navigation behavior', () => {
  it('should navigate to the correct path', async () => {
    const { user } = setupSidebarNavigationItem();

    expect(screen.getByRole('link', { name: 'Test Item' })).toBeInTheDocument();

    await user.click(screen.getByRole('link', { name: 'Test Item' }));
    expect(screen.getByText('/second')).toBeInTheDocument();
    expect(screen.queryByRole('link', { name: 'Test Item' })).not.toBeInTheDocument();
  });
});

it("shouldn't have any a11y violation", async () => {
  const { container } = setupSidebarNavigationItem({ Icon: IconBranch });
  await expect(container).toHaveNoA11yViolations();
});

function setupSidebarNavigationItem(props: Partial<SidebarNavigationItemProps> = {}) {
  return renderWithMemoryRouter(
    <SidebarNavigationItem to="/second" {...props}>
      Test Item
    </SidebarNavigationItem>,
  );
}
