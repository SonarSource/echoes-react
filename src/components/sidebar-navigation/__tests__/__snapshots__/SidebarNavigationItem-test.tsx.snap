// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`should render with an icon 1`] = `
.emotion-0 {
  all: unset;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: var(--echoes-dimension-space-100);
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  box-sizing: border-box;
  height: var(--echoes-dimension-height-800);
  padding: var(--echoes-dimension-space-100);
  border-radius: var(--echoes-border-radius-400);
  color: var(--echoes-color-text-default);
  font: var(--echoes-typography-text-default-regular);
  cursor: pointer;
}

.emotion-0:hover {
  background-color: var(--echoes-color-background-selected-weak-default);
}

.emotion-0:focus-visible {
  outline: var(--echoes-color-focus-default) solid var(--echoes-focus-border-width-default);
  outline-offset: var(--echoes-focus-border-offset-default);
  border-radius: var(--echoes-border-radius-400);
}

.emotion-0:active,
.emotion-0.active {
  background-color: var(--echoes-color-surface-hover);
  color: var(--echoes-color-text-accent);
  font: var(--echoes-typography-text-default-semi-bold);
}

.emotion-0:active:hover,
.emotion-0.active:hover {
  background-color: var(--echoes-color-background-selected-weak-hover);
}

.emotion-2 {
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  font-size: calc(1em + 4px);
  font-style: normal;
  font-weight: normal;
  height: calc(2em - 16px);
  line-height: calc(2em - 16px);
  text-align: center;
  vertical-align: bottom;
  width: calc(2em - 16px);
  font-family: 'Echoes';
}

.emotion-4 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<a
  class="emotion-0 emotion-1"
  href="/second"
>
  <span
    aria-hidden="true"
    class="emotion-2 emotion-3"
    css="[object Object]"
  >
    !
  </span>
  <span
    class="emotion-4 emotion-5"
  >
    Test Item
  </span>
</a>
`;
